Source/Events.Handling/Builder/ConventionEventHandlerBuilder.cs

```diff
@@ -0,0 +1,319 @@
+            }
+
+            if (!MethodHasNoExtraParameters(method))
+            {
+                warnings.Add($"Event handler method {method} on event handler {_eventHandlerType} needs to only have two parameters where the first is the event to handle and the second is {typeof(EventContext)}");
+            }
+
+            if (MethodReturnsAsyncVoid(method) || (!MethodReturnsVoid(method) && !MethodReturnsTask(method)))
+            {
+                warnings.Add($"Event handler method {method} on event handler {_eventHandlerType} needs to return either {typeof(void)} or {typeof(Task)}");
+            }
+
+            if (warnings.Count == 0) return true;
+            buildWarnings = new EventHandlerBuildWarnings(eventHandlerId, warnings);
+            return false;
+        }
+
+        void AddUntypedHandleSignatureToBuilder(MethodInfo method, EventType eventType, EventHandlerMethodsBuilder builder)
+        {
+            var eventHandlerSignatureType = method.ReturnType == typeof(Task) ? typeof(EventHandlerSignature) : typeof(VoidEventHandlerSignature);
+            var eventHandlerSignature = Delegate.CreateDelegate(eventHandlerSignatureType, _eventHandlerInstance, method);
+            var builderHandleMethod = typeof(EventHandlerMethodsBuilder)
+                                        .GetMethod(nameof(EventHandlerMethodsBuilder.Handle), new[] { typeof(EventType), eventHandlerSignatureType });
+            if (eventHandlerSignature is EventHandlerSignature signature) builder.Handle(eventType, signature);
+            else if (eventHandlerSignature is VoidEventHandlerSignature voidSignature) builder.Handle(eventType, voidSignature);
+        }
+
+        void AddTypedHandleSignatureToBuilder(Type eventParameterType, MethodInfo method, EventHandlerMethodsBuilder builder)
+        {
+            var eventHandlerSignatureGenericTypeDefinition = method.ReturnType == typeof(Task) ?
+                                                typeof(TypedEventHandlerSignature<>)
+                                                : typeof(VoidTypedEventHandlerSignature<>);
+            var eventHandlerSignatureType = eventHandlerSignatureGenericTypeDefinition.MakeGenericType(eventParameterType);
+            var eventHandlerSignature = Delegate.CreateDelegate(eventHandlerSignatureType, _eventHandlerInstance, method);
+            var builderHandleMethod = typeof(EventHandlerMethodsBuilder)
+                                        .GetMethods()
+                                        .Where(_ => _.IsGenericMethod && _.Name == nameof(EventHandlerMethodsBuilder.Handle))
+                                        .Single(_ => _.GetParameters()[0].ParameterType.GetGenericTypeDefinition() == eventHandlerSignatureGenericTypeDefinition)
+                                        .MakeGenericMethod(eventParameterType);
+            builderHandleMethod.Invoke(builder, new[] { eventHandlerSignature });
```

Comment:
Wæt is going on here? This might be something that I invented a long time ago, but either way it is not really all that easy to make sense of. Seems like maybe we could make this simpler by making some new implementations of `IEventHandlerMethod` here maybe?

Reply:
Maybe we should make this class without using the `EventHandlerMethodsBuilder`?

Reply:
This should be resolved by latest commits